### DISCUSSION TITLE
No field.onChange types?

### DISCUSSION THREAD

**Discussion by josias-r** [react-hook-form/react-hook-form#9192] [Q&A]
After a long time of reporting [an issue with setValue](https://github.com/react-hook-form/react-hook-form/issues/7602), where I learned I should be using `field.onChange` instead of `setValue` in the first place, I finally got around adjusting our codebase to this.

Now that I started implementing the onChange, I realized though, that unlike `setValue`, `onChange` is hardly typed:

```
ControllerRenderProps<...> = {
   ...
   onChange: (...event: any[]) => void;
}
```

With `setValue("a-number-field", "string")` the hook-form would throw a proper TS error, if the field is not compatible with type `string`.
While using `field.onChange("whatever")` does not throw any TS errors of course, since it's arguments are type `any`: `...event: any[]`

What is the reason behind this?

**Comment by bluebill1049** [comment]
because it could be an event object.

**Reply by josias-r** [reply]
@bluebill1049 Okay. I had the freedom in our codebase to overwrite the changeHandler types, since we have a wrapper for your hook.

But for general use, this would be great, if it were fully typed.
From looking at [getEventValue.ts](https://github.com/react-hook-form/react-hook-form/blob/master/src/logic/getEventValue.ts), it seems like the `onChange` arguments could be of type: `Event | FieldValue`. Also no need for the array type, since it looks like only the first argument (event or value) is ever used.
